### ISSUE TITLE
Implement hand-over to Terra by reference

### ISSUE THREAD

**Issue by hannes-ucsc** [DataBiosphere/azul#4121]
**The description is out of date, the updated specification can be found here: https://github.com/DataBiosphere/azul/issues/4121#issuecomment-1272056479**

In order to facilitate a script that populates a post-handover workspace with verbatim metadata from the source snapshots of the entities included in the handover, we need to include, for each entity in the PFB manifest (aka row in a post-handover workspace table), the following items:

- the UUID of the source snapshot the entity was indexed from 
- the `datarepo_row_id` of the entity
- the value of the `source_datarepo_row_ids` column of the entity

The  `source_datarepo_row_ids` is a list of of colon separated values like `ws_sequencing:5b4e73b3-4d43-4388-a7d7-6f3eb2211e05`. Each value references a row in another table (in the same snapshot) that the entity (the row in the referencing table) was subset from. The referencing table has a well-formed schema, the referenced table's schema does not, and is specific to the consortium or even the study.

**Comment by theathorn** [comment]
Is this needed for the stakeholder demo?

**Comment by melainalegaspi** [comment]
@hannes-ucsc to add description motivating the need for this ticket.

**Comment by hannes-ucsc** [comment]
Reducing estimate because I think we can implement this using the current PFB hand-over implementation that's backed by the aggregation index.

**Comment by mbaumann-broad** [comment]
@theathorn cc: @hannes-ucsc 
> Is this needed for the stakeholder demo?

Yes. Recently (9/20/22) Rachel identified the need to have the Terra user workspace data structure/content closely resemble the original Terra data workspace created by the data submitter.
This is important/necessary to be acceptable to the AnVIL program sponsors.
This feature is needed to enable that.

**Comment by mbaumann-broad** [comment]
@hannes-ucsc 
To help ensure we agree on the result of this change (including #4482 and #4483), following a PFB handoff to a Terra workspace, an excerpt of the `donor` table would look like the following.
The `source_datarepo_row_ids` column will already exist within the snapshot itself and is to be passed through to the PFB, as described in #4482.
The `source_datarepo_snapshot` column needs to be insert and Azul indexing time and included in the PFB, as described in #4483.

![donor_table_excerpt](https://user-images.githubusercontent.com/84099036/193088893-28734137-2bb7-45bf-a8ab-b734c46a157e.jpg)

**Comment by theathorn** [comment]
Added to next Tuesday's agenda.

**Comment by melainalegaspi** [comment]
@hannes-ucsc to review.

**Comment by hannes-ucsc** [comment]
@mbaumann-broad, so far I have not heard a convincing argument as to why we need to complicate the snapshot reference to use the composite, versioned syntax as you propose. The consumer can, and should always look-up the snapshot name by the snaphot's ID. It is a design smell to decorate foreign keys (and that's what the snapshot reference is) with redundant information that is evident from the bare foreign key itself.

**Comment by mbaumann-broad** [comment]
@hannes-ucsc , yes, including only the snapshot UUID is _technically_ sufficient, purely from an implementation perspective.

And, based on clarification of the scope/meaning of the `version` field with the TDR team, I agree that it is not needed here.

Yet, user/researcher usability of the data in the Terra data tables is also important, and that is the reason for the inclusion of the snapshot name. As currently defined, the AnVIL snapshot naming convention provides useful provenance information at a glance. This is particularly useful when a researcher is performing cross-program or cross-dataset/study analysis, where a given table contains data from multiple sources. The snapshot name is immutable, so the snapshot name and snapshot UUID will remain consistent. 

The alternative is to have a Terra data table column that provides only a UUID and provides no direct value to the user. I do not generally expect researchers to go to the trouble to look up the name for a snapshot given only the UUID, nor do I see why they should need to make that additional effort.

I expect it may take a few more lines of code in Azul, perhaps in a few places, to support this. Yet, from a researcher usability perspective, I think it is worth it.

Finally, as a potential additional benefit ...
Although snapshots used to populate user workspaces should never be deleted, if that happened, the snapshot UUID alone would be useless, while the snapshot name would continue to provide a durable indicator of provenance that could help in updating the workspace to a newer snapshot that is available/supported.

To recap, the format of the `datarepo_source_snapshot` column values should be:
_snapshot_name:snapshot_uuid_

**Comment by hannes-ucsc** [comment]
I think this is conflating *provenance* with *readability*, maybe that's why I didn't understand your previous motivation for this. Provenance is the ability to determine the origin of something. The UUID gives us that. I hope that's uncontested at this point.

Let me ask this: Why would a researcher want to know the snapshot name? What information can be glanced from the snapshot name, information that isn't already encapsulated in the `dataset` table? Please note that each row in each table in the handover workspace will have a FK to the dataset table, where all info about the dataset is already represented. It requires a simple join to get that. 

Assuming that we come up with a good answer to that question, I might be persuaded to include the snapshot name. However, I truly dislike the proprietary `:` syntax. One relational schema design principle is to have column values to be atomic. So if we include the name, we'd have to include it as a separate column.

Please also note that the snapshot name is not sufficient to form a fully-qualified BQ table name. You'd need the Google Cloud project name, too. So even if we include the name, it would still require a round-trip to TDR to build the table name for any BQ queries your script might be doing.

**Comment by melainalegaspi** [comment]
@hannes-ucsc to wrap up discussion with the Broad and record the outcome here.

**Comment by mbaumann-broad** [comment]
Thanks @melainalegaspi. This ball is in my court now, and I have more information to add, yet it is consistent with the recent meeting decision for the value to be the snapshot id only. It does not affect/change that decision.
I'll add this soon so this is clearly settled/closed, yet need to get some other things done first.

**Comment by hannes-ucsc** [comment]
Updated specification:

In order to facilitate a script that populates a post-handover workspace with verbatim metadata from the source snapshots of the entities included in the handover, we need to include, for each entity in the PFB manifest (each entity corresponds to a row in a post-handover workspace table), the following items:

- the UUID of the source snapshot the entity was indexed from, in a column called `source_datarepo_snapshot_id`
- the `datarepo_row_id` of the entity, in a column of the same name
- the value of the `source_datarepo_row_ids` column of the the entity, in a column of the same name

The  `source_datarepo_row_ids` column is a JSON array of JSON objects. Each object is of the form 

```json
{
    "table_name": "ws_sequencing", 
    "datarepo_row_id": "5b4e73b3-4d43-4388-a7d7-6f3eb2211e05"
}
```

Each list element references a row in another table that the entity (the row in the referencing table) was subset from. The referencing table has a well-formed schema, the referenced table's schema typically does not, and is specific to the consortium or even the study. The referenced table resides in the snapshot whose ID is `source_datarepo_snapshot_id`.

**Comment by hannes-ucsc** [comment]
@mbaumann-broad please review the [updated specification](https://github.com/DataBiosphere/azul/issues/4121#issuecomment-1272056479) and let us know if it works for you.

**Comment by mbaumann-broad** [comment]
@hannes-ucsc, the newly proposed `datarepo_row_id` column consists of a single row id, whereas the `source_datarepo_row_ids` value may contain multiple values. For the current/known use cases, a single `datarepo_row_id` will be sufficient. What are your thoughts on how `datarepo_row_id` would be handled in the case of multiple values for `source_datarepo_row_ids` in the future?

**Comment by mbaumann-broad** [comment]
@hannes-ucsc, oh, I think you are saying that `datarepo_row_id` is the id of the (Findability Subset) entity itself, and not otherwise related to the `source_datarepo_row_ids`, correct?

**Comment by hannes-ucsc** [comment]
Great, thank you. We'll schedule this for the current sprint.

**Comment by hannes-ucsc** [comment]
This will be demoed as part of #4588.

**Comment by noah-aviel-dove** [comment]
@hannes-ucsc: "In order to be able to include three new columns in workspace tables other than `files`, they need to be aggregated using a set. We'll set the size limit to 100 and log a warning when we are dropping items due to that limit. That way we'll know to increase the limit. This is still just a stopgap solution. I could imagine certain graphs that would easily require a set limit of 100,000 or more, which would be impractical. A complete solution requires backing the PFB manifest by contribution indices rather than just aggregate ones (https://github.com/DataBiosphere/azul/issues/4311)."
